lib/backend/dynamo/dynamodbbk.go

```diff
@@ -617,45 +683,60 @@ const (
 
 // DeleteRange deletes range of items with keys between startKey and endKey
 func (b *Backend) DeleteRange(ctx context.Context, startKey, endKey backend.Key) error {
-	if startKey.IsZero() {
-		return trace.BadParameter("missing parameter startKey")
-	}
-	if endKey.IsZero() {
-		return trace.BadParameter("missing parameter endKey")
-	}
-	// keep fetching and deleting until no records left,
-	// keep the very large limit, just in case if someone else
-	// keeps adding records
-	for i := 0; i < backend.DefaultRangeLimit/100; i++ {
-		result, err := b.getRecords(ctx, prependPrefix(startKey), prependPrefix(endKey), batchOperationItemsLimit, nil)
+	// Attempt to pull all existing items and delete them in batches
+	// in accordance with the BatchWriteItem limits. There is a hard
+	// cap on the total number of items that can be deleted in a single
+	// DeleteRange call to avoid racing with additional records being added.
+	const maxDeletionOperations = backend.DefaultRangeLimit / 100 / batchOperationItemsLimit
+	requests := make([]types.WriteRequest, 0, batchOperationItemsLimit)
+	var deletions int
+	for item, err := range b.Items(ctx, backend.IterateParams{StartKey: startKey, EndKey: endKey}) {
 		if err != nil {
 			return trace.Wrap(err)
 		}
-		if len(result.records) == 0 {
-			return nil
+
+		if deletions >= maxDeletionOperations {
+			break
 		}
```

Comment:
No need to pull one more item from the iterator, move this right after the `deletions++`.